Lesson_Materials/Lecture_20_More_SYCL_Features/index.html

```diff
@@ -0,0 +1,355 @@
+						<div class="col" data-markdown>
+```
+q.parallel_for([=](sycl::item&lt;1&gt; it) {
+  sycl::atomic_ref&lt;T,
+          sycl::memory_order_relaxed, 
+          sycl::memory_scope_device&gt;
+            (a[0])
+            .fetch_add(it.get_global_linear_id());
+});
+```
+						</div>
+					</div>
+				</section>
+				<!--Slide 9 -->
+				<section>
+					<div class="hbox" data-markdown>
+#### Atomic Operations
+					</div>
+					<div class="container">
+						<div class="col" data-markdown>
+![SYCL](./combine3.png "SYCL")
+						</div>
+						<div class="col" data-markdown>
+```
+q.parallel_for([=](sycl::item&lt;1&gt; it) {
+  sycl::atomic_ref&lt;T,
+          sycl::memory_order_relaxed, 
+          sycl::memory_scope_device&gt;
+            (a[0])
+            .fetch_add(it.get_global_linear_id());
+});
+```
+* Using atomics, values can be combined without data races.
+* `fetch_add`, `fetch_sub`, `fetch_max`, `fetch_min` are some of the ways we can
+  combine values atomically.
+* `fetch_and` and `fetch_or` can also be used for
+  integral types
+* Please see the SYCL Specification for more details.
+						</div>
+					</div>
+				</section>
+				<!--Slide 10 -->
+				<section>
+					<div class="hbox" data-markdown>
+#### Atomic Operations
+					</div>
+					<div class="container">
+						<div class="col">
+							<pre><code>
+q.parallel_for([=](sycl::item&lt;1&gt; it) {
+  sycl::atomic_ref&lt;T,
+          sycl::memory_order_relaxed,
+          sycl::memory_scope_<span class="highlight">work_group</span>,
+          <span class="highlight">sycl::access::address_space::local_space</span>&gt;
+            (a[0])
+            .fetch_add(it.get_global_linear_id());
+});
+							</code></pre>
+						</div>
+						<div class="col" data-markdown>
+* We can also specify the memory space of `a[0]`.
+* If `a[0]` is in local memory, we should expect a speedup for using the local
+	memory atomic over the default atomic (which uses the generic address space).
+						</div>
+					</div>
+				</section>
+				<!--Slide 11 -->
+				<section>
+					<div class="hbox" data-markdown>
+#### Group Algorithms
+					</div>
+					<div class="container" data-markdown>
+* SYCL provides group algorithms which perform common operations over a single
+	workgroup.
+* `reduce` algorithms perform fast work group reduction operation for some op
+	such as `plus`, `max`, etc.
+* `any_of`, `all_of`, `none_of` as well as the `joint_*` counterparts perform
+	some predicate checking and return the same value to all items in a work 
+	group.
+* `permute_group_by_xor` permutes values among work items in a work group,
+	according to a provided mask.
+* `inclusive_scan` and `exclusive_scan`. For a scan of elements: `[x0,...,xn]`
+	the ith result of an exclusive scan is the combination of the elements
+	[x0,...,x{i-1}] and the ith result of an inclusive scan is the combination of
+	elements [x0,...,xi] using some binary op.
+* Please see the SYCL specification for more details.
```

Comment:
You could also add that group algorithms can operate on different group scopes and that they have the requirement that all work-items in that group scope must call the function in convergent control flow.

Reply:
Change made